force-app/main/default/classes/DMLOperations.cls

```diff
@@ -180,28 +252,56 @@ public with sharing class DMLOperations {
 	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
 	 */
 	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
-		// Loop through each Contact
 
-		// Extract the Account Name from Contact's LastName
+        String nameOfAccount;
+        for(Contact contact : contactsToUpsert){
+            nameOfAccount = contact.LastName; 
+        }
 
 		// Get the account based on the name or create one if it doesn't exist
+        List<Account> accounts = [SELECT Id FROM Account WHERE Name = :nameOfAccount];
+        if(accounts.isEmpty()){
+            Account newAccount = new Account(Name = nameOfAccount);
+            Database.upsert(newAccount);
+        } 
 
 		// Link the Contact to its Account using the Account Id
+        String accountId;
+        for(Contact contact : contactsToUpsert){
+            for(Account account : ([SELECT Id FROM Account WHERE Name = :contact.LastName])){
```

Comment:
Normally directly querying in your loop is a great way to cut down on extra variables and code. However, this for loop is already nested in another loop so its going against best practices.